File: server/routes/workspaces.py
"""Safe local workspace alternatives to a native folder picker.

中文说明：工作区 API：获取主目录、列出已登记工作区、创建默认工作区，
以及安全地登记用户输入的工作区路径（替代原生文件夹选择器）。
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, Request
from pydantic import BaseModel

from server.errors import APIError
from server.services.workspace_service import WorkspaceService

router = APIRouter(tags=["workspaces"])


class SelectWorkspaceRequest(BaseModel):
    cwd: str


def get_workspace_service(request: Request) -> WorkspaceService:
    return request.app.state.workspace_service


@router.get("/api/home")
async def get_home(service: WorkspaceService = Depends(get_workspace_service)) -> dict[str, str]:
    """返回受控工作区父目录。"""
    return {"home": str(service.workspace_parent)}


@router.get("/api/workspaces")
async def list_workspaces(service: WorkspaceService = Depends(get_workspace_service)) -> dict:
    """列出当前允许的所有工作区根目录。"""
    return {"workspaces": [str(root) for root in service.roots()]}


@router.post("/api/default-cwd")
async def create_default_workspace(
    service: WorkspaceService = Depends(get_workspace_service),
) -> dict[str, str]:
    """在受控父目录下创建默认工作区（pi-cwd-日期）。"""
    try:
        return {"cwd": str(service.create_default())}
    except OSError as exception:
        raise APIError(500, "workspace_create_failed", str(exception)) from exception


@router.post("/api/workspaces/select")
async def select_workspace(
    body: SelectWorkspaceRequest,
    service: WorkspaceService = Depends(get_workspace_service),
) -> dict[str, str]:
    """登记一个已有目录为工作区；必须位于父目录或已登记工作区之下。"""
    try:
        return {"cwd": str(service.select(body.cwd))}
    except ValueError as exception:
        raise APIError(400, "invalid_workspace", str(exception)) from exception
    except PermissionError as exception:
        raise APIError(403, "workspace_not_allowed", str(exception)) from exception

File: node-pi/server/src/routes/workspaces.ts
/**
 * 工作区（Workspace）路由，供现有 Vue 的目录选择器使用。
 *
 * 中文说明：接口字段与 FastAPI 后端保持一致，Vue 不需要区分当前使用哪个后端。
 * 核心能力：查询用户主目录、列出已登记的工作区、创建默认工作区、
 * 弹系统目录选择器（Windows）、以及登记/选择某个工作区。
 *
 * Fastify 概念：这些路由都注册在根路径下（未使用 prefix），
 * 路径在插件内部写全（/api/home 等）；handler 里 return 的对象自动 JSON 化。
 */

import type { FastifyPluginAsync } from "fastify";

import { ApiError } from "../errors.js";
import { WorkspaceService } from "../services/workspace-service.js";

/** 插件选项：工作区服务。 */
export interface WorkspaceRouteOptions {
  service: WorkspaceService;
}

export const workspaceRoutes: FastifyPluginAsync<WorkspaceRouteOptions> = async (app, options) => {
  // GET /api/home —— 返回默认工作区父目录（配置的 workspaceParent，通常是用户主目录）。
  app.get("/api/home", async () => ({ home: options.service.parent }));

  // GET /api/workspaces —— 列出所有已登记且仍然存在的本地工作区目录。
  app.get("/api/workspaces", async () => ({ workspaces: await options.service.roots() }));

  // POST /api/default-cwd —— 在父目录下按日期创建默认工作区（如 pi-cwd-20250701），
  // 并自动登记。首次使用、还没选择目录时前端会调用它。
  app.post("/api/default-cwd", async () => ({ cwd: await options.service.createDefault() }));

  // POST /api/workspaces/pick —— 弹出 Windows 原生"选择文件夹"对话框，
  // 用户取消时返回 { cwd: null }。
  app.post("/api/workspaces/pick", async () => ({ cwd: await options.service.pickDirectory() ?? null }));

  // POST /api/workspaces/select —— 登记一个已存在的本地目录为工作区。
  // body: { cwd: "<绝对路径>" }，路径不存在或不是目录会抛 400。
  app.post("/api/workspaces/select", async (request) => {
    const body = request.body;
    if (body === null || typeof body !== "object" || Array.isArray(body) || typeof (body as { cwd?: unknown }).cwd !== "string") {
      throw new ApiError(422, "validation_error", "cwd must be a string");
    }
    return { cwd: await options.service.select((body as { cwd: string }).cwd) };
  });
};
